man/makeinfo.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.4.
.TH TEXI2ANY "1" "May 2017" "texi2any 6.3.90" "User Commands"
.SH NAME
texi2any \- translate Texinfo documents
.SH SYNOPSIS
.B makeinfo
[\fI\,OPTION\/\fR]... \fI\,TEXINFO-FILE\/\fR...
.SH DESCRIPTION
Translate Texinfo source documentation to various other formats, by default
Info files suitable for reading online with Emacs or standalone GNU Info.
.PP
This program is commonly installed as both `makeinfo' and `texi2any';
the behavior is identical, and does not depend on the installed name.
.SS "General options:"
.TP
\fB\-\-document\-language\fR=\fI\,STR\/\fR locale to use in translating Texinfo keywords
for the output document (default C).
.TP
\fB\-\-error\-limit\fR=\fI\,NUM\/\fR
quit after NUM errors (default 100).
.TP
\fB\-\-force\fR
preserve output even if errors.
.TP
\fB\-\-help\fR
display this help and exit.
.TP
\fB\-\-no\-validate\fR
suppress node cross\-reference validation.
.TP
\fB\-\-no\-warn\fR
suppress warnings (but not errors).
.TP
\fB\-\-conf\-dir\fR=\fI\,DIR\/\fR
search also for initialization files in DIR.
.TP
\fB\-\-init\-file\fR=\fI\,FILE\/\fR
load FILE to modify the default behavior.
.TP
\fB\-c\fR, \fB\-\-set\-customization\-variable\fR VAR=VAL
set customization variable VAR
to value VAL.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
explain what is being done.
.TP
\fB\-\-version\fR
display version information and exit.
.SS "Output format selection (default is to produce Info):"
.TP
\fB\-\-docbook\fR
output Docbook XML rather than Info.
.TP
\fB\-\-html\fR
output HTML rather than Info.
.TP
\fB\-\-plaintext\fR
output plain text rather than Info.
.TP
\fB\-\-xml\fR
output Texinfo XML rather than Info.
.TP
\fB\-\-dvi\fR, \fB\-\-dvipdf\fR, \fB\-\-ps\fR, \fB\-\-pdf\fR
call texi2dvi to generate given output,
after checking validity of TEXINFO\-FILE.
.SS "General output options:"
.TP
\fB\-E\fR, \fB\-\-macro\-expand\fR=\fI\,FILE\/\fR
output macro\-expanded source to FILE,
ignoring any @setfilename.
.TP
\fB\-\-no\-headers\fR
suppress node separators, Node: lines, and menus
from Info output (thus producing plain text)
or from HTML (thus producing shorter output).
Also, if producing Info, write to
standard output by default.
.TP
\fB\-\-no\-split\fR
suppress any splitting of the output;
generate only one output file.
.TP
\fB\-\-[no\-]number\-sections\fR
output chapter and sectioning numbers;
default is on.
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,DEST\/\fR
output to DEST.
With split output, create DEST as a directory
.TP
and put the output files there.
With non\-split output, if DEST is already
.TP
a directory or ends with a /,
put the output file there.
.IP
Otherwise, DEST names the output file.
.SS "Options for Info and plain text:"
.TP
\fB\-\-disable\-encoding\fR
do not output accented and special characters
in Info output based on @documentencoding.
.TP
\fB\-\-enable\-encoding\fR
override \fB\-\-disable\-encoding\fR (default).
.TP
\fB\-\-fill\-column\fR=\fI\,NUM\/\fR
break Info lines at NUM characters (default 72).
.TP
\fB\-\-footnote\-style\fR=\fI\,STYLE\/\fR
output footnotes in Info according to STYLE:
`separate' to put them in their own node;
`end' to put them at the end of the node, in
which they are defined (this is the default).
.TP
\fB\-\-paragraph\-indent\fR=\fI\,VAL\/\fR
indent Info paragraphs by VAL spaces (default 3).
If VAL is `none', do not indent; if VAL is
`asis', preserve existing indentation.
.TP
\fB\-\-split\-size\fR=\fI\,NUM\/\fR
split Info files at size NUM (default 300000).
.SS "Options for HTML:"
.TP
\fB\-\-css\-include\fR=\fI\,FILE\/\fR
include FILE in HTML <style> output;
read stdin if FILE is \-.
.TP
\fB\-\-css\-ref\fR=\fI\,URL\/\fR
generate CSS reference to URL.
.TP
\fB\-\-internal\-links\fR=\fI\,FILE\/\fR
produce list of internal links in FILE.
.TP
\fB\-\-split\fR=\fI\,SPLIT\/\fR
split at SPLIT, where SPLIT may be `chapter',
`section' or `node'.
.TP
\fB\-\-transliterate\-file\-names\fR
use file names in ASCII transliteration.
.TP
\fB\-\-node\-files\fR
produce redirection files for nodes and
anchors; default is set only if split.
.SS "Options for XML and Docbook:"
.TP
\fB\-\-output\-indent\fR=\fI\,VAL\/\fR
does nothing, retained for compatibility.
.SS "Options for DVI/PS/PDF:"
.TP
\fB\-\-Xopt\fR=\fI\,OPT\/\fR
pass OPT to texi2dvi; can be repeated.
.SS "Input file options:"
.TP
\fB\-\-commands\-in\-node\-names\fR
does nothing, retained for compatibility.
.TP
\fB\-D\fR VAR
define the variable VAR, as with @set.
.TP
\fB\-D\fR 'VAR VAL'
define VAR to VAL (one shell argument).
.TP
\fB\-I\fR DIR
append DIR to the @include search path.
.TP
\fB\-P\fR DIR
prepend DIR to the @include search path.
.TP
\fB\-U\fR VAR
undefine the variable VAR, as with @clear.
.SS "Conditional processing in input:"
.TP
\fB\-\-ifdocbook\fR
process @ifdocbook and @docbook even if
not generating Docbook.
.TP
\fB\-\-ifhtml\fR
process @ifhtml and @html even if not generating HTML.
.TP
\fB\-\-ifinfo\fR
process @ifinfo even if not generating Info.
.TP
\fB\-\-ifplaintext\fR
process @ifplaintext even if not generating plain text.
.TP
\fB\-\-iftex\fR
process @iftex and @tex.
.TP
\fB\-\-ifxml\fR
process @ifxml and @xml.
.TP
\fB\-\-no\-ifdocbook\fR
do not process @ifdocbook and @docbook text.
.TP
\fB\-\-no\-ifhtml\fR
do not process @ifhtml and @html text.
.TP
\fB\-\-no\-ifinfo\fR
do not process @ifinfo text.
.TP
\fB\-\-no\-ifplaintext\fR
do not process @ifplaintext text.
.TP
\fB\-\-no\-iftex\fR
do not process @iftex and @tex text.
.TP
\fB\-\-no\-ifxml\fR
do not process @ifxml and @xml text.
.P
Also, for the \fB\-\-no\-ifFORMAT\fR options, do process @ifnotFORMAT text.
.P
The defaults for the @if... conditionals depend on the output format:
if generating Docbook, \fB\-\-ifdocbook\fR is on and the others are off;
if generating HTML, \fB\-\-ifhtml\fR is on and the others are off;
if generating Info, \fB\-\-ifinfo\fR is on and the others are off;
if generating plain text, \fB\-\-ifplaintext\fR is on and the others are off;
if generating XML, \fB\-\-ifxml\fR is on and the others are off.
.SH EXAMPLES
.TP
makeinfo foo.texi
write Info to foo's @setfilename
.TP
makeinfo \-\-html foo.texi
write HTML to @setfilename
.TP
makeinfo \-\-xml foo.texi
write Texinfo XML to @setfilename
.TP
makeinfo \-\-docbook foo.texi
write Docbook XML to @setfilename
.TP
makeinfo \-\-plaintext foo.texi
write plain text to standard output
.TP
makeinfo \-\-pdf foo.texi
write PDF using texi2dvi
.TP
makeinfo \-\-html \-\-no\-headers foo.texi
write html without node lines, menus
.TP
makeinfo \-\-number\-sections foo.texi
write Info with numbered sections
.TP
makeinfo \-\-no\-split foo.texi
write one Info file however big
.SH "REPORTING BUGS"
Email bug reports to bug\-texinfo@gnu.org,
general questions and discussion to help\-texinfo@gnu.org.
.br
Texinfo home page: http://www.gnu.org/software/texinfo/
.SH COPYRIGHT
Copyright \(co 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B texi2any
is maintained as a Texinfo manual.  If the
.B info
and
.B texi2any
programs are properly installed at your site, the command
.IP
.B info texi2any
.PP
should give you access to the complete manual.


NEWS
$Id$
This NEWS file records noteworthy changes, very tersely.
See the manual for detailed information.

  Copyright 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001,
  2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012,
  2013, 2014, 2015, 2016, 2017 Free Software Foundation, Inc.

  Copying and distribution of this file, with or without modification,
  are permitted in any medium without royalty provided the copyright
  notice and this notice are preserved.

------------------------------------------------------------------------------

* info:
  . do not fall back to a man page if following a cross-reference in an
    info file failed

* texi2any
  . `INLINE_INSERTCOPYING' removed as a customization variable


6.4 (23 June 2017)
* texi2any:
  . for HTML output, place section names before the manual in page
    titles, instead of after them, so it is easier to distinguish pages
    if titles are truncated
  . starting points for ordered lists beginning with 0 or a letter of
    the alphabet are output as attributes on the <ol> tag, as was
    case for Texinfo 4.13 and earlier
  . a bit faster
  . some discrepancies in paragraph formatting between Perl extension
    modules and interpreted Perl modules have been fixed
  . `MACRO_BODY_IGNORES_LEADING_SPACE' customization variable removed,
    and `indent_menu_descriptions' is no longer a possible value for
    `TREE_TRANSFORMATIONS' (as these features did not work as
    documented)

* info:
  . the `up-line' and `down-line' commands now are like the other
    scrolling commands and are no longer confined to a single node
    (depending on the value of `scroll-behaviour')
  . supplying the --all option with --index-search displays a list
    of matching index entries
  . the style variables like `link-style' can now be set while info
    is running
  . display bug fixed where color could be turned off prematurely
  . several other bugs fixed
  . better portability in test suite
  . do not fall back to showing the dir node if a manual isn't found
  . Do not attempt any kind of conversion of CR LF line endings,
    except on MS-DOS/Windows, when it is done unconditionally.  (This 
    replaces a more complicated approach, where this conversion would 
    take place if there was a problem finding a node in a file.)  The 
    main effect of this change is that Info files with CR LF line 
    endings, which would have been produced on MS-DOS/Windows with old 
    versions of makeinfo, are only supported on such operating systems.
  . a few of the key bindings under --vi-keys have been changed for
    consistency or to match the documentation

* texinfo.tex
  . a DVI file with a single empty page can be output again, which
    restores the behavior from Texinfo 6.0 and earlier

* Distribution
  . autoconf 2.69, automake 1.15, gettext 0.19.8



6.3 (10 September 2016)

* Language:
  . The commands `@setcontentsaftertitlepage' and
    `@setshortcontentsaftertitlepage' have been removed.
  . @-commands are no longer supported within `@errormsg'.

* texinfo.tex:
  . For a couple of characters (opening and closing braces), use glyphs
    from the standard TeX math fonts instead of using EC fonts which are 
    less likely to be installed.
  . Use of user-defined macros in the text of an index entry is more
    reliable when the text contains Texinfo @-commands.
  . @synindex and @syncodeindex have been fixed (broken in the last
    release).
  . Support added for native UTF-8 support with XeTeX and LuaTeX.
  . Support of PDF output with XeTeX improved.
  . You can use a new file doc/texinfo-ja.tex for Texinfo documents in
    Japanese.  doc/short-sample-ja.tex is a sample document.  New 
    support file doc/txi-ja.tex for Japanese.

* texi2any:
  . Fix handling of compiler options when building Perl extension modules.

* texi2dvi:
  . Can now process files whose absolute paths contain space characters,
    as long a relative path to the file is given.  Better support of 
    files with unusual characters in their names.
  . No longer exits prematurely in some circumstances (due to the script
    running under "set -e").
  . Bug fixed which made the `--command' and `--tidy' options
    incompatible.

* info:
  . Handling of "invalid" value in infokey file fixed.

6.2 (withdrawn)



6.1 (06 February 2016)

* Language:
  . You can now omit the @menu from nodes with other nodes below them in
    the document structure.  If you use "@validatemenus off" near the 
    start of a Texinfo file, makeinfo will, where needed, create a menu 
    for nodes lacking one given explicitly.
  . An @setfilename line is no longer required at the start of a
    Texinfo file.  (Be aware, though, that some other tools may require 
    it, for example Automake.)
  . For processing with TeX, a comma is automatically provided following
    a cross-reference command (such as @xref) when needed to separate
    the page number from following text, so you don't need to add one
    yourself.  See the `Parts of a Cross Reference' node in the manual
    for details.  Behavior when followed by punctuation, as always
    recommended previously, is unchanged.

* texi2any:
  . Some Perl modules have been rewritten in C to increase speed.
    If Perl extensions can be created, they are used by default; 
    otherwise the pure Perl implementations are still used.
    Disable at build time with "configure --disable-perl-xs".  The 
    environment variable TEXINFO_XS controls how they are used by 
    texi2any.
  . Quotation marks are left out for node names and index entries in
    Info output where they would have been produced by commands such
    as @file or @option.
  . New customization variable INFO_SPECIAL_CHARS_QUOTE to allow use of
    a quoting mechanism for problematic constructs in Info output, for
    example node names containing colons or commas.
  . Commands like @heading are affected by @lowersections and @raisesections

texinfo/texi2html/examples/makeinfo.pm
# -*-perl-*-
#+##############################################################################
#
# makeinfo.pm: output html in the makeinfo style.
#
# This is obsoleted and irrelevant now, in 2010 (though still used in some 
# tests) since a makeinfo-like default is used when called as makeinfo or 
# texi2any.
#
#    Copyright (C) 2003, 2005, 2008 Free Software Foundation, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License,
# or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Originally written by Patrice Dumas.
#
#-##############################################################################

#html_default_load();
t2h_default_load_format('html');

set_default('TOP_FILE', 'index.html');

set_default('SHOW_MENU', 1);

set_default('USE_NODES', 1);

set_default('SPLIT', 'node');

set_default('NODE_FILES', 1);

set_default('USE_SETFILENAME', 1);

set_default('HEADERS', 1);

set_default('SHORT_REF', 1);

set_default('NODE_NAME_IN_MENU', 1);

set_default('NODE_NAME_IN_INDEX', 0);

set_default('AVOID_MENU_REDUNDANCY', 0);

set_default('SPLIT_INDEX', 0);

$MENU_SYMBOL = '*';

set_default('WORDS_IN_PAGE', 200);

set_default('footnotestyle', 'end');

set_default('USE_ACCESSKEY', 1);

set_default('USE_REL_REV', 1);

set_default('USE_LINKS', 1);

set_default('INLINE_CONTENTS', 1);

set_default('OVERVIEW_LINK_TO_TOC', 1);

if (get_conf('HEADER_IN_TABLE'))
{
    @SECTION_BUTTONS =
    (
     \"node: ", \$Texi2HTML::NODE{This}, [ 'NodeNext', \$Texi2HTML::NODE{NodeNext} ] , \",", [ 'NodePrev', \$Texi2HTML::NODE{NodePrev} ], \",", [ 'NodeUp', \$Texi2HTML::NODE{NodeUp} ], 'NodeForward'
    );

    @TOP_BUTTONS = @SECTION_BUTTONS;
    @MISC_BUTTONS = @SECTION_BUTTONS;
}

push @command_handler_process, \&makeinfo_like_initialize_variables;

my $makekinfo_like_footnote_absolute_number;
my %makekinfo_like_paragraph_in_footnote_nr;

sub makeinfo_like_initialize_variables()
{
   $makekinfo_like_footnote_absolute_number = 0;
   %makekinfo_like_paragraph_in_footnote_nr = ();
}

my $default_init_out = $init_out;
$init_out = \&makeinfo_like_init_out;

sub makeinfo_like_init_out()
{
   &$default_init_out();
   $NAVIGATION_TEXT{'NodeForward'} = ' &gt; ';
}

# FIXME: images not in paragraphs are in <div class="block-image">


$foot_line_and_ref  = \&makeinfo_like_foot_line_and_ref;
$foot_section       = \&makeinfo_like_foot_lines;
$paragraph          = \&makeinfo_like_paragraph;
$index_summary      = \&makeinfo_like_index_summary;
$summary_letter     = \&makeinfo_like_summary_letter;
$index_entry        = \&makeinfo_like_index_entry;
$index_letter       = \&makeinfo_like_index_letter;
$print_index        = \&makeinfo_like_print_index;
$listoffloats       = \&makeinfo_like_listoffloats;
$listoffloats_entry = \&makeinfo_like_listoffloats_entry;
$menu_link          = \&makeinfo_like_menu_link;
$menu_description   = \&makeinfo_like_menu_description;
$menu_comment       = \&makeinfo_like_menu_comment;
$menu               = \&makeinfo_like_menu;

sub makeinfo_like_foot_line_and_ref($$$$$$$$)
{
    my $foot_num = shift;
    my $relative_num = shift;
    my $footid = shift;
    my $docid = shift;
    my $from_file = shift;
    my $footnote_file = shift;
    my $lines = shift;
    my $state = shift;
    
    $makekinfo_like_footnote_absolute_number++;

    # this is a bit obscure, this allows to add an anchor only if formatted
    # as part of the document.
    $docid = '' if ($state->{'outside_document'} or $state->{'multiple_pass'});

    if ($from_file eq $footnote_file)
    { 
        $from_file = $footnote_file = '';
    }

    my $foot_anchor = "<sup>" . &$anchor($docid, "$footnote_file#$footid", $relative_num) . "</sup>";
    $foot_anchor = &$anchor($docid, "$footnote_file#$footid", "($relative_num)") if ($state->{'preformatted'});

    return ($lines, $foot_anchor);
}

sub makeinfo_like_foot_lines($)
{
    my $lines = shift;
    unshift @$lines, "<div class=\"footnote\">\n<hr>\n<h4>". gdt('Footnotes') ."</h4>\n";
    push (@$lines, "</div>\n");
    return $lines;
}

sub makeinfo_like_paragraph ($$$$$$$$$$$$$)
{
    my $text = shift;
    my $align = shift;
    my $indent = shift;
    my $paragraph_command = shift;
    my $paragraph_command_formatted = shift;
    my $paragraph_number = shift;
    my $format = shift;
    my $item_nr = shift;
    my $enumerate_style = shift;
    my $number = shift;
    my $command_stack_at_end = shift;
    my $command_stack_at_begin = shift;
    my $state = shift;
#print STDERR "format: $format\n" if (defined($format));
#print STDERR "paragraph @$command_stack_at_end; @$command_stack_at_begin\n";
    #$paragraph_command_formatted = '' if (!defined($paragraph_command_formatted) or
    #      exists($special_list_commands{$format}->{$paragraph_command}));
    return '' if ($text =~ /^\s*$/);
    if (defined($paragraph_number) and defined($$paragraph_number))
    {
         $$paragraph_number++;
         return $text  if (($format eq 'itemize' or $format eq 'enumerate') and
            ($$paragraph_number == 1));
    }
    my $open = '<p';
    if ($align)
    {
        $open .= " align=\"$paragraph_style{$align}\"";
    }

    my $footnote_text = '';
    if (defined($command_stack_at_begin->[0]) and $command_stack_at_begin->[0] eq 'footnote')
    {
        my $state = $Texi2HTML::THISDOC{'state'};
        $makekinfo_like_paragraph_in_footnote_nr{$makekinfo_like_footnote_absolute_number}++;
        if ($makekinfo_like_paragraph_in_footnote_nr{$makekinfo_like_footnote_absolute_number} <= 1)
        {
           $open.=' class="footnote"';
           my $document_file = $state->{'footnote_document_file'};
           if ($document_file eq $state->{'footnote_footnote_file'})
           {
               $document_file = '';
           }
           my $docid = $state->{'footnote_place_id'};
           my $doc_state = $state->{'footnote_document_state'};
           $docid = '' if ($doc_state->{'outside_document'} or $doc_state->{'multiple_pass'});
           my $foot_label = &$anchor($state->{'footnote_footnote_id'}, 
                 $document_file . "#$state->{'footnote_place_id'}", 
                 "$state->{'footnote_number_in_page'}");
           $footnote_text = "<small>[${foot_label}]</small> ";
        }
    }

    $text = $footnote_text.$text;

    my $top_stack = '';
    $top_stack = $command_stack_at_begin->[-1] if (scalar (@$command_stack_at_begin));
    if ($top_stack eq 'multitable')
# FIXME
# paragraph_number is not by cell, so it is not possible to know if it
# is the first paragraph in cell. Also currently there is no hook for a
# cell beginning which means that it is not possible to reset the cell 
# paragraph count at a cell beginning
# and (!defined($paragraph_number) or !defined($$paragraph_number) or $$paragraph_number == 1))
    {
       return $text;
    }

    return $open.'>'.$text.'</p>';
}


sub makeinfo_like_index_summary
{
    return '';
}

sub makeinfo_like_summary_letter
{
    return '';
}

sub makeinfo_like_index_entry($$$$$$$$$$)
{
    my $text_href = shift;
    my $entry = shift;
    my $element_href = shift;
    my $element_text = shift;
    my $entry_file = shift;
    my $current_element_file = shift;
    my $entry_target = shift;
    my $entry_element_target = shift;
    my $in_region_not_in_output = shift;
    my $index_entry_ref = shift;

    return '' if ($entry !~ /\S/);
    my $element = $index_entry_ref->{'real_element'};
    if (defined($element))
    {
       my $element_set = 0;
       if (get_conf('NODE_NAME_IN_INDEX'))
       {
           if ($element->{'node'})
           {
               $element_set = 1;
           }
           elsif ($element->{'with_node'})
           {
               $element = $element->{'with_node'};
               $element_set = 1;
           }
       }
       elsif (defined(get_conf('NODE_NAME_IN_INDEX')))
       {
           if (!$element->{'node'})
           {
               $element_set = 1;
           }
           elsif ($element->{'with_section'})
           {
               $element = $element->{'with_section'};
               $element_set = 1;
           }
       }
       if ($element_set)
       {
           $element_href = main::href($element, $Texi2HTML::THIS_ELEMENT->{'file'},
                 $Texi2HTML::THISDOC{'line_nr'});
           $element_text = $element->{'text'};
       }
    }

    
    return '<li>' . $entry . ': ' .  &$anchor('', $element_href, $element_text)
    . "</li>\n";
}

sub makeinfo_like_index_letter($$$)
{
     my $letter = shift;
     my $id = shift;
     my $text = shift;
     return $text;
}

sub makeinfo_like_print_index($$)
{
    my $text = shift;
    my $name = shift;
    return '' if (!defined($text));
    return "<ul class=\"index-$name\" compact>\n" . $text . "</ul>\n";
}

sub makeinfo_like_listoffloats_entry($$$$)
{
    my $style_texi = shift;
    my $float = shift;
    my $float_style = shift;
    my $caption = shift;
    my $href = shift;

    if ($caption ne '')
    {
        return '<li>' . gdt('{href}: {caption}', { 'href' => &$anchor('', $href, $float_style), 'caption' => $caption }) . "</li>\n";
    }
    else
    {
        return '<li>' . &$anchor('', $href, $float_style) . "</li>\n";
    }
}

sub makeinfo_like_listoffloats($$$)
{
    my $style_texi = shift;
    my $style = shift;
    my $float_entries = shift;

    my $result = "<ul class=\"listoffloats\">\n" ;
    foreach my $float_entry (@$float_entries)
    {
         $result .= $float_entry;
    }
    return $result . "</ul>\n";
}

# a whole menu
#
# argument:
# the whole menu text (entries and menu comments)
#
# argument:
# whole menu text.
sub makeinfo_like_menu($)
{
    my $text = shift;
    $menu_entry_index=0;
    if ($text =~ /\S/)
    {
        return "<ul>\n" 
        . $text . "</ul>\n";
    }
}

# formats a menu entry link pointing to a node or section 
#
# arguments:
# the entry text
# the state, a hash reference holding informations about the context, with a 
#     usefull entry, 'preformatted', true if we are in a preformatted format
#     (a format keeping space between words). In that case a function
#     of the main program, main::do_preformatted($text, $state) might 
#     be used to format the text with the current format style.
# href is optionnal. It is the reference to the section or the node anchor
#     which should be used to make the link (typically it is the argument 
#     of a href= attribute in a <a> element).
sub makeinfo_like_menu_link($$$$$$$)
{
    my $entry = shift;
    my $state = shift;
    my $href = shift;
    my $node = shift;
    my $title = shift;
    my $ending = shift;
    my $has_title = shift;
    $title = '' unless ($has_title);
    if (($entry eq '') or get_conf('NODE_NAME_IN_MENU') or $state->{'preformatted'})
    {
        
        $title .= ':' if ($title ne '');
        $entry = "$title$node";
    }
    $menu_entry_index++;
    my $accesskey;
    $accesskey = "accesskey=\"$menu_entry_index\"" if (get_conf('USE_ACCESSKEY') and ($menu_entry_index < 10));
    $entry = &$anchor ('', $href, $entry, $accesskey) if (defined($href));
#    if ($state->{'preformatted'})
#    {
#        return '<li>' . main::do_preformatted($entry . $ending, $state);
#    }
    return "<li> $entry";
}

# formats a menu entry description, ie the text appearing after the node
# specification in a menu entry an spanning until there is another
# menu entry, an empty line or some text at the very beginning of the line
# (we consider that text at the beginning of the line begins a menu comment) 
#
# arguments:
# the description text
# the state. See menu_entry.
sub makeinfo_like_menu_description($$)
{
    my $text = shift;
    my $state = shift;
#    if ($state->{'preformatted'})
#    {
#        return main::do_preformatted($text, $state) . '</td></tr>';
#    }
    return "$text</li>\n";
}

1;
